Move numa.{h, cc} to c10/util

jerryzh168 opened the pull request:
Summary: att

Differential Revision: D13205604

ezyang commented on c10/util/numa.h:
```diff
@@ -0,0 +1,21 @@
+
+namespace c10 {
+
+C10_API bool IsNUMAEnabled();
```
If we're going to move this to c10, let's document everything? :)

jerryzh168 replied:
Sure I'll add some brief documents, @ilia-cher could you write a more detailed one if there are more things to explain when this is landed?

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,132 @@
+#include "c10/util/numa.h"
+
+C10_DEFINE_bool(caffe2_cpu_numa_enabled, false, "Use NUMA whenever possible.");
```
@dzhulgakov @ilia-cher Do we actually need this flag? I'd much rather if we just always have NUMA enabled if it's available, saves a fiddly configuration flag.

ilia-cher replied:
without it we will treat legacy protobufs as having all CPU ops pinned to NUMA node 0.

dzhulgakov replied:
hm, so right now it's used only in some benchmarks from what I can tell. How do you plan to eventually roll out it widely?

jerryzh168 replied:
@ilia-cher so simply removing the flag in all places would work?

ilia-cher replied:
yes, there're a few places (allocator) that will result in extra calls but these calls should be no-ops (since numa node id is -1 by default), make sure to rerun tests though

jerryzh168 replied:
what are the tests? @ilia-cher

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,132 @@
+
+  CAFFE_ENFORCE(
+      numa_node_id <= numa_max_node(),
+      "NUMA node id " + c10::to_string(numa_node_id) + " is unavailable");
```
`to_string` call here isn't necessary; if we use the `AT_CHECK` macro it will automatically stringify.

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,132 @@
+      "NUMA node id " + c10::to_string(numa_node_id) + " is unavailable");
+
+  auto bm = numa_allocate_nodemask();
+  numa_bitmask_clearall(bm);
```
`numa_allocate_nodemask` is specified in the documentation to return a zero-filled bitmask. Thus, the clearall call is not necessary.

jerryzh168 replied:
cc @ilia-cher

ilia-cher replied:
good catch

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,132 @@
+  int numa_node = -1;
+  CAFFE_ENFORCE(
+      get_mempolicy(
+          &numa_node, NULL, 0, (void*)ptr, MPOL_F_NODE | MPOL_F_ADDR) == 0,
```
Cast to `void*` is not necessary in C++; this conversion will be done implicitly

jerryzh168 replied:
looks like we need `const_cast`?

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,132 @@
+  CAFFE_ENFORCE(
+      get_mempolicy(
+          &numa_node, NULL, 0, (void*)ptr, MPOL_F_NODE | MPOL_F_ADDR) == 0,
+      "Unable to get memory policy");
```
This is a bad way to test for error. On failure, `get_mempolicy` is specified to set `errno`. We should report the errno appropriately if there is a failure.

ilia-cher replied:
"On success, get_mempolicy() returns 0; on error, -1 is returned and errno is set to indicate the error."

ilia-cher replied:
(not arguing we should not report errno though)

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,132 @@
+  return numa_node;
+}
+
+int GetNumNUMANodes() {
```
Can we return an `optional<int>` here instead of using -1 to indicate that it's not enabled? Since we're in C++ we don't have to make C-style compromises.

ezyang replied:
(This applies to other sites too.)

jerryzh168 replied:
will change in next diff

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,132 @@
+  }
+  CAFFE_ENFORCE(ptr);
+
+  size_t page_start_ptr = (((size_t)ptr) & ~(getpagesize() - 1));
```
1. This is using a C-style cast, we should not use C-style casts.
2. The standard doesn't guarantee that casting a pointer to `size_t` to be well defined. Use `uintptr_t` instead.

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,137 @@
+    return;
+  }
+  if (!IsNUMAEnabled()) {
+    VLOG(1) << "NUMA is not enabled";
```
@ilia-cher Do you know what the motivation behind the logging here? Was it just to be screamy when someone accidentally deploys caffe2 in fbcode without numa support? I can't see most people actually finding this log message being useful.

ilia-cher replied:
it's a log message to inform that NUMABind requested (numa_node_id >= 0) but is not performed, that would result in errors or low perf

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,137 @@
+int GetNUMANode(const void* ptr) {
+  if (!IsNUMAEnabled()) {
+    VLOG(1) << "NUMA is not enabled";
+    return {};
```
Err... I don't think you wanted to return {}, if your return type is an int.

jerryzh168 replied:
we have a followup diff that changes the type to optional<int> and we're returning c10::nullopt there.

ezyang replied:
Yes, but if you want to stack your diffs, each intermediate diff has to be correct...

jerryzh168 replied:
this means 0 for int? is it?

ezyang replied:
Yes, and before it was -1

jerryzh168 replied:
@ezyang actually I just copied the code, so it's 0 before as well.  what should we return here?
cc @ilia-cher

ezyang commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,137 @@
+  uintptr_t page_start_ptr = ((reinterpret_cast<uintptr_t>(ptr)) & ~(getpagesize() - 1));
+  ptrdiff_t offset = reinterpret_cast<uintptr_t>(ptr) - page_start_ptr;
+  // Avoid extra dynamic allocation and NUMA api calls
+  AT_ASSERT(numa_node_id >= 0 && (unsigned)numa_node_id < sizeof(unsigned long) * 8);
```
C-style cast here...

ezyang reviewed (changes requested):
returning {} is almost certainly wrong

jerryzh168 commented on c10/CMakeLists.txt:
```diff
@@ -41,6 +41,28 @@ if (${COMPILER_SUPPORTS_HIDDEN_VISIBILITY})
 endif()
 
+# ---[ Check for NUMA support
+if (USE_NUMA)
```
@ilia-cher can we just move this check from caffe2 to c10 like this? and do we need to do anything for `USE_NUMA`?

ilia-cher replied:
cc. @orionr @pietern  to check CMake changes

dzhulgakov replied:
cc @Yangqing

orionr replied:
I would remove C10_IS_NUMA_AVAILABLE and C10_DISABLE_NUMA and simply update the USE_NUMA in cmake/Dependencies.cmake to do these CHECK_CXX_SOURCE_COMPILES  include checks if needed.

jerryzh168 commented on c10/util/numa.cpp:
```diff
@@ -0,0 +1,137 @@
+#include "c10/util/numa.h"
+
+#if USE_NUMA && defined(__linux__) && !defined(C10_DISABLE_NUMA) && C10_MOBILE == 0
```
here I added `USE_NUMA`

dzhulgakov replied:
how exactly does it work? C++ includes and cmake variables are two totally different namespaces

orionr replied:
You should update c10/macros/cmake_macros.h.in to get C10_USE_NUMA to get passed through instead of leaving it in caffe2.

dzhulgakov commented on caffe2/core/macros.h.in:
```diff
@@ -43,7 +43,7 @@ static_assert(
 #cmakedefine CAFFE2_USE_NVTX
 #cmakedefine CAFFE2_USE_TRT
-#cmakedefine CAFFE2_DISABLE_NUMA
+#cmakedefine C10_DISABLE_NUMA
```
wait, you probably want to move necessary flags here - https://github.com/pytorch/pytorch/blob/master/c10/macros/cmake_macros.h.in

and please, please test that CMake flags actually work as expected :-)

orionr replied:
Yup - you should really move this.

orionr reviewed (changes requested):
Some CMake changes needed. Thanks!

ezyang commented on caffe2/core/macros.h.in:
```diff
@@ -81,5 +80,4 @@ static_assert(
   {"USE_MKLDNN", "${CAFFE2_USE_MKLDNN}"}, \
   {"USE_NVTX", "${CAFFE2_USE_NVTX}"}, \
   {"USE_TRT", "${CAFFE2_USE_TRT}"}, \
-  {"DISABLE_NUMA", "${CAFFE2_DISABLE_NUMA}"}, \
```
Wow! Is there really no where else that uses `CAFFE2_DISABLE_NUMA`? I'm impressed

jerryzh168 replied:
yeah that is true

ezyang reviewed (approved):
The VLOGs are still a stick up my craw, but I guess we can go with this.

jerryzh168 committed:
Move numa.{h, cc} to c10/util (#14393)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/14393

att

Reviewed By: ezyang

Differential Revision: D13205604

fbshipit-source-id: aa001af9bd04784877f0364bd9fabf65afc0af68
